### PR TITLE
Added range check of error codes

### PR DESCRIPTION
**Pull request by dmatousek**
The PR adds named constants for min and max values of error, warning, and info message codes to be used in backends. Also, the `ErrorCatalog::add` method now expects a code type so that it can perform range check. E.g.:

`ErrorCatalog::getCatalog().add<ErrorMessage::MessageType::Info, INFO_USER_CODE>("user-code");`

If the range check fails, a bug is triggered.

In addition, a new info code INFO_PROGRESS is added.

### CODE REVIEW

File: lib/error_catalog.h
Comment (vlstill) [reply lib/error_catalog.h:109]: It would require actually calling `add` in compile-time context. However map::emplace is not constexpr (not even in C++23). Or it would need to be a template, somewhat like this:

```cpp
template<MessageType type, int errCode>
bool add(const char *name, bool forceReplace = flase) {
   static_assert(type !=  MessageType::Error ||
            (errorCode >= ErrorType::ERR_MIN_BACKEND && errorCode < ErrorType::ERR_MAX),
            "Error code out of range");
   ...
}
```

That would work too (I believe). It would be called like this:

```cpp
errorCatalog.add<MessageType::Error, Err::ERR_FOO>("foo");
```

Which is not super-nice, but I'd say it is worth the added safety.

For this to work, the constants would need to be declared in the `.h` file. Probably as

```cpp
static constexpr int ERR_MIN_BACKEND = ...;
```

Comment (fruffy) [reply lib/error_catalog.h:109]: I'll let you decide whether you consider it worth implementing. I think it'd be nice, since runtime `BUG`s are a persistence source of annoyance.

**Review by fruffy — APPROVED** [review]

File: lib/error_catalog.h
Comment (dmatousek) [reply lib/error_catalog.h]: Thanks for your suggestions, they really make sense!

File: lib/error_catalog.h
Comment (fruffy) [inline lib/error_catalog.h:112]: Neat, does this work `errorcode` or should we make explicit that it is a `constexpr`? Same for all the error codes.

Comment (dmatousek) [reply lib/error_catalog.h:112]: Sorry, what do you mean?

Comment (fruffy) [reply lib/error_catalog.h:112]: All the integers are `const` not `constexpr`. Does this work with `static_assert`? Or does the compiler convert them automatically to compile time constants?

Comment (dmatousek) [reply lib/error_catalog.h:112]: Yes, it works. The compilation fails with:
```
lib/error_catalog.h:112:50: error: static assertion failed
  112 |         static_assert(type != MessageType::Error ||
      |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
  113 |             (errorCode >= ErrorType::ERR_MIN_BACKEND && errorCode <= ErrorType::ERR_MAX));
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
I believe that for an integral type `const` implies `consexpr`.

**Review by fruffy — APPROVED** [review]

**Comment by dmatousek** [comment]
What do I have to do to be able to merge p4c PRs?

**Comment by fruffy** [comment]
I think we need to add you as a committer. Working on that. I can merge it for you.

**Comment by dmatousek** [comment]
Thank you!

**Comment by fruffy** [comment]
There should be a person at Intel that can add you to p4lang, is that not the case anymore?

**Pull request merged**
